### PR TITLE
Hide ads on amp pages if shouldHideAdverts is true

### PR DESCRIPTION
**Pull request by TBonnin**
## What does this change?

Hides ads on amp pages if shouldHideAdverts is set to true for this article
Fix for https://github.com/guardian/frontend/issues/12025
## Request for comment

@NataliaLKB @johnduffell

### CODE REVIEW

File: article/app/views/package.scala [hunk 1]
```diff
@@ -36,7 +38,7 @@ object MainCleaner {
 object BodyCleaner {
   def apply(article: Article, html: String, amp: Boolean)(implicit request: RequestHeader) = {
     implicit val edition = Edition(request)
-    val cleaners = List(
+    val cleaners = ListBuffer(
```
Comment (TBonnin) [inline article/app/views/package.scala]: Making this list mutable. Since it is only a local variable it shouldn't be a issue.

File: article/app/views/package.scala [hunk 2]
```diff
@@ -59,13 +61,16 @@ object BodyCleaner {
...
-    withJsoup(BulletCleaner(html))((if (amp) ampOnlyCleaners else nonAmpCleaners) ++ cleaners :_*)
+
+    if (amp) {
+      cleaners += AmpEmbedCleaner(article)
+      if(!article.content.shouldHideAdverts) {
+        cleaners += AmpAdCleaner(edition, request.uri, article)
+      }
+    } else {
+      cleaners += VideoEmbedCleaner(article)
+    }
```
Comment (TBonnin) [inline article/app/views/package.scala]: If there is a more idiomatic way of creating this list I am happy to do it.

**Comment by jbreckmckye** [comment]
Can content from the children's book site appear on AMP? Those pages are often not flagged as sensitive, but cannot show advertising - ASA regulations prohibit it on anything that might be deemed childrens' content.

**Comment by TBonnin** [comment]
@jbreckmckye Thanks for pointing this out. Indeed this logic happens there for the website. https://github.com/guardian/frontend/blob/f3941b51b79b63abf92fc9c53dd2f7af8201b694/static/src/javascripts/projects/common/modules/commercial/commercial-feature-policies.js#L24  

Duplicating this kind of logic for amp and non-amp will for sure end up in oversights and bugs in the future and I don't think we can always count on attentive reviewers like you to catch this kind of mistmatches.
This might be a very stupid question but why is CAPI not returning `shouldHideAdverts=false` for all `/childrens-book-site/...` content?

**Comment by jbreckmckye** [comment]
@TBonnin - I don't really know how CAPI works or the taxonomy of backend content. There must presumably be some reason we haven't done this, but I'm not sure who to ask about it.

**Comment by NataliaLKB** [comment]
@TBonnin @jbreckmckye maybe ask @mchv or @cb372

**Comment by mchv** [comment]
@jbreckmckye let's have a chat about it when you will be available

File: article/app/views/package.scala (hunk 1 above)
Comment (johnduffell) [inline article/app/views/package.scala]: if you're going to have a mutable list, it's better if it's local, but the same arguments about being able to reason about your code apply regardless.  I'd recommend keeping it immutable.
Having said all that, there may be a better (but still functional) way of building up the list as a logic for whats in it becomes more complicated.

Comment (TBonnin) [inline article/app/views/package.scala]: The problem with using an immutable list in this case is you'll have to make a bunch of copies and temporary vals, or use weird filtering. Unless I am missing something...

**Comment by TBonnin** [comment]
@mchv @jbreckmckye have you chatted about this?

**Comment by jbreckmckye** [comment]
@TBonnin We have. It does make sense; the property is an aspect of presentation, and that's not CAPI's concern. So any rationalization needs to happen in the rendering tier.

File: article/app/views/package.scala [hunk 3]
```diff
@@ -58,14 +64,11 @@ object BodyCleaner {
...
-    )
-    val ampOnlyCleaners = List(
-      AmpAdCleaner(edition, request.uri, article),
-      AmpEmbedCleaner(article)
-    )
-    withJsoup(BulletCleaner(html))((if (amp) ampOnlyCleaners else nonAmpCleaners) ++ cleaners :_*)
+    ) ++
+      ListIf(!amp)(VideoEmbedCleaner(article)) ++
+      ListIf(amp)(AmpEmbedCleaner(article)) ++
+      ListIf(amp && shouldShowAds)(AmpAdCleaner(edition, request.uri, article))
```
Comment (TBonnin) [inline article/app/views/package.scala]: @johnduffell @dominickendrick you might like that better :)

File: article/app/views/package.scala [hunk 4]
```diff
@@ -36,6 +38,10 @@ object MainCleaner {
 object BodyCleaner {
   def apply(article: Article, html: String, amp: Boolean)(implicit request: RequestHeader) = {
     implicit val edition = Edition(request)
+
+    val shouldShowAds = !article.content.shouldHideAdverts && article.metadata.section != "childrens-books-site"
```
Comment (TBonnin) [inline article/app/views/package.scala]: Got it @jbreckmckye 
I guess the section check should be centralised somewhere instead of having this check scattered in multiple places in scala/javascript code. However I feel this goes behind the scope of this PR and could be the objective of another pull request

Comment (JonNorman) [inline article/app/views/package.scala]: We have [`commercial-features.js`](https://github.com/guardian/frontend/blob/master/static/src/javascripts/projects/common/modules/commercial/commercial-features.js) which does codify some of these checks (see `var sensitiveContent`), which you could plug in.

File: article/app/views/package.scala (hunk 3 above)
Comment (johnduffell) [inline article/app/views/package.scala]: clever! looks good :yum: thanks

**Comment by johnduffell** [comment]
:+1: I'm happy anyway, improved on the previous, but feel free to deal with the duplication too

File: article/app/views/package.scala [hunk 5]
```diff
@@ -8,6 +8,8 @@ import play.api.mvc.RequestHeader
 import views.support._
 import views.support.cleaner.{AmpEmbedCleaner, VideoEmbedCleaner, CmpParamCleaner, AmpAdCleaner}
 
+import scala.collection.mutable.ListBuffer
```
Comment (johnduffell) [inline article/app/views/package.scala]: dn't forget to remove the unneeded import

Comment (TBonnin) [inline article/app/views/package.scala]: Good catch. I will remove. Thx

File: article/app/views/package.scala (hunk 4 above)
Comment (TBonnin) [inline article/app/views/package.scala]: @JonNorman That's happening client side but on amp page ads are included server side during page generation.

**Comment by NataliaLKB** [comment]
:+1: Thanks for adding the bit about Children's books!
